circuitbreaker/hystrix_test.go

```diff
@@ -0,0 +1,116 @@
+package circuitbreaker_test
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/afex/hystrix-go/hystrix"
+	"github.com/go-kit/kit/circuitbreaker"
+	"github.com/go-kit/kit/endpoint"
+	"golang.org/x/net/context"
+)
+
+func TestHystrixCircuitBreakerOpen(t *testing.T) {
+	var (
+		thru       = 0
+		myError    = error(nil)
+		ratio      = 0.04
+		primeWith  = hystrix.DefaultVolumeThreshold * 2
+		shouldPass = func(failed int) bool { return (float64(failed) / float64(primeWith+failed)) <= ratio }
+		extraTries = 10
+	)
+
+	// configure hystrix
+	hystrix.ConfigureCommand("myEndpoint", hystrix.CommandConfig{
+		ErrorPercentThreshold: 5,
+	})
+
+	var e endpoint.Endpoint
+	e = func(context.Context, interface{}) (interface{}, error) { thru++; return struct{}{}, myError }
+	e = circuitbreaker.Hystrix("myEndpoint")(e)
+
+	// prime
+	for i := 0; i < primeWith; i++ {
+		if _, err := e(context.Background(), struct{}{}); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Now we start throwing errors.
+	myError = errors.New(":(")
+
+	// The first few should get thru.
+	var letThru int
+	for i := 0; shouldPass(i); i++ { // off-by-one
+		letThru++
+		if _, err := e(context.Background(), struct{}{}); err != myError {
+			t.Fatalf("want %v, have %v", myError, err)
+		}
+	}
+
+	// But the rest should be blocked by an open circuit.
+	for i := 1; i <= extraTries; i++ {
+		if _, err := e(context.Background(), struct{}{}); err != circuitbreaker.ErrCircuitBreakerOpen {
+			t.Errorf("with request #%d, want %v, have %v", primeWith+letThru+i, circuitbreaker.ErrCircuitBreakerOpen, err)
+		}
+	}
+
+	// Confirm the rest didn't get through.
+	if want, have := primeWith+letThru, thru; want != have {
+		t.Errorf("want %d, have %d", want, have)
+	}
+}
```

Comment:
We should factor a lot of this out to a general circuitbreaker test driver, but I'll do that afterwards.